### PR TITLE
[IRN-5328] [BpkDrawer] supports non-padded content

### PR DESCRIPTION
**Pull request by steviehailey-skyscanner**
Add a new prop: `padded` to allow the content to be completely controlled by the consumer without using the `contentClassName` prop.

Context:
* We have a design to implement where the content is full width and has a non-white background colour. We can acheive this using the `contentClassName` prop to remove some padding being added by default but there is a risk of CSS non-determinism issues reverting our padding override.
* Instead we add a new option to optionally disable the default padding so there is no need to override backpack styles. 
* Slack discussion: https://skyscanner.slack.com/archives/C0JHPDSSU/p1726658223807479

Change is backwards compatible with existing behaviour but allows consumers to move away from solution which risks CSS non-determinism issues.

Changes
* BpkDrawer has new prop `padded` (default `true`)
* BpkDrawerContent child component has new prop `padded` (no default as only consumed via BpkDrawer)
  * BpkDrawerContent converted to TypeScript
  * (_Looked to also convert BpkDrawer to TS but seems best to convert to functional component at the same time and want to avoid adding that scope to this PR_) 

Remember to include the following changes:

- [x] Ensure the PR title includes the name of the component you are changing so it's clear in the release notes for consumers of the changes in the version e.g `[KOA-123][BpkButton] Updating the colour`
- [x] `README.md` (If you have created a new component)
- [ ] Component `README.md`
- [x] Tests
- [x] Accessibility tests
    - The following checks were performed:
        - [x] Ability to navigate using a [keyboard only](https://webaim.org/techniques/keyboard/)
        - [x] Zoom functionality ([Deque University explanation](https://dequeuniversity.com/checklists/web/text)):
            - [x] The page SHOULD be functional AND readable when only the text is magnified to 200% of its initial size
            - [x] Pages must reflow as zoom increases up to 400% so that content continues to be presented in only one column i.e. Content MUST NOT require scrolling in two directions (both vertically and horizontally)
        - [x] Ability to navigate using a [screen reader only](https://webaim.org/articles/screenreader_testing/)
- [x] Storybook examples created/updated
- [x] For breaking changes or deprecating components/properties, migration guides added to the description of the PR. If the guide has large changes, consider creating a new Markdown page inside the component's docs folder and link it here

### CODE REVIEW

File: examples/bpk-component-drawer/examples.js [hunk 1]
```diff
@@ -262,10 +262,28 @@ const WithFullHeightContentExample = () => (
   </DrawerContainer>
 );
 
+const WithNonPaddedContentExample = () => (
```
Comment (steviehailey-skyscanner) [inline examples/bpk-component-drawer/examples.js:265]: New test scenario for added prop based on surrounding cases

File: examples/bpk-component-drawer/stories.js [hunk 2]
```diff
@@ -25,6 +24,7 @@ import {
   CloseButtonTextExample,
   WithVisuallyHiddenTitleExample,
   WithFullHeightContentExample,
+  WithNonPaddedContentExample,
```
Comment (steviehailey-skyscanner) [inline examples/bpk-component-drawer/stories.js:27]: New story to demo use of the feature

File: packages/bpk-component-drawer/src/BpkDrawer.js [hunk 3]
```diff
@@ -118,6 +119,7 @@ BpkDrawer.defaultProps = {
   closeLabel: null,
   closeText: null,
   hideTitle: false,
+  padded: true,
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/BpkDrawer.js:122]: New option - default to `true` for backwards compatibility. Consumers can opt-out for new non-padded behaviour.

File: examples/bpk-component-drawer/examples.module.scss [hunk 4]
```diff
@@ -30,3 +30,15 @@
 .bpk-drawer-button {
   margin: auto auto 0;
 }
+
+.bpk-drawer-content-full-width {
```
Comment (steviehailey-skyscanner) [inline examples/bpk-component-drawer/examples.module.scss:34]: Style for new story

File: packages/bpk-component-drawer/src/BpkDrawerContent.tsx [hunk 5]
```diff
@@ -16,16 +16,14 @@
  * limitations under the License.
  */
 
-/* @flow strict */
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/BpkDrawerContent.tsx:19]: Converted to TS

File: packages/bpk-component-drawer/src/BpkDrawerContent.tsx [hunk 6]
```diff
@@ -35,41 +33,42 @@ const getClassName = cssModules(STYLES);
...
-    onClose,
-    onCloseAnimationComplete,
-    title,
-    ...rest
-  } = props;
+const BpkDrawerContent = ({
+  children,
+  className,
+  closeLabel,
+  closeOnScrimClick = true, // Unused from withScrim scrim HOC
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/BpkDrawerContent.tsx:57]: Can still opt out of this but click to close kept as default

File: packages/bpk-component-drawer/src/BpkDrawerContent.tsx [hunk 7]
```diff
@@ -16,16 +16,14 @@
...
-
-import PropTypes from 'prop-types';
-import type { Node } from 'react';
-
+// @ts-expect-error Untyped import. See `decisions/imports-ts-suppressions.md`.
 import Transition from 'react-transition-group/Transition';
 
 import { animations } from '@skyscanner/bpk-foundations-web/tokens/base.es6';
 
+// @ts-expect-error Untyped import. See `decisions/imports-ts-suppressions.md`.
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/BpkDrawerContent.tsx:24]: Standard comment used when imported module is not TS yet...

File: packages/bpk-component-drawer/src/BpkDrawerContent.tsx [hunk 8]
```diff
@@ -35,41 +33,42 @@ const getClassName = cssModules(STYLES);
...
+  dialogRef: () => React.RefObject<HTMLElement>,
+  onCloseAnimationComplete: () => void,
+  onClose: () => () => void
   id: string,
   title: string,
-  className: ?string,
-  contentClassName: ?string,
-  closeLabel: string,
-  closeText: ?string,
+  className?: string,
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/BpkDrawerContent.tsx:41]: Optional fields...

File: packages/bpk-component-drawer/src/BpkDrawerContent.tsx [hunk 9]
```diff
@@ -83,6 +82,10 @@ const BpkDrawerContent = (props: Props) => {
     headerClassNames.push(getClassName('bpk-drawer__heading--visually-hidden'));
   }
 
+  if (padded) {
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/BpkDrawerContent.tsx:85]: New logic to allow avoiding padding

File: packages/bpk-component-drawer/src/BpkDrawerContent.tsx [hunk 10]
```diff
@@ -101,11 +104,10 @@ const BpkDrawerContent = (props: Props) => {
       in={isDrawerShown}
       onExited={onCloseAnimationComplete}
     >
-      {(status) => (
-        // $FlowFixMe[cannot-spread-inexact] - inexact rest. See decisions/flowfixme.md
+      {(status: string) => (
         <section
           id={id}
-          tabIndex="-1"
+          tabIndex={-1}
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/BpkDrawerContent.tsx:110]: TS fix

File: packages/bpk-component-drawer/src/__snapshots__/BpkDrawerContent-test.js.snap [hunk 11]
```diff
@@ -165,6 +165,59 @@ exports[`BpkDrawerContent should render correctly when it has a contentClassName
...
+            >
+              <path
+                d="M5.587 3.467a1.5 1.5 0 1 0-2.194 2.046q.036.038.074.074l6.438 6.44-6.44 6.44a1.5 1.5 0 0 0 2.122 2.12l6.44-6.438 6.44 6.44a1.5 1.5 0 0 0 2.12-2.122l-6.438-6.44 6.44-6.44a1.5 1.5 0 0 0-2.122-2.12l-6.44 6.438-6.44-6.44z"
+              />
+            </svg>
+          </span>
+        </button>
+      </div>
+    </header>
+    <div
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/__snapshots__/BpkDrawerContent-test.js.snap:212]: No padded class for new scenario

File: packages/bpk-component-drawer/src/__snapshots__/BpkDrawerContent-test.js.snap [hunk 12]
```diff
@@ -165,6 +165,59 @@ exports[`BpkDrawerContent should render correctly when it has a contentClassName
...
+      class="bpk-drawer__content bpk-drawer__content--padded"
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/__snapshots__/BpkDrawerContent-test.js.snap:213]: Padded class being added

File: packages/bpk-component-drawer/src/__snapshots__/BpkDrawer-test.js.snap [hunk 13]
```diff
@@ -61,7 +61,7 @@ exports[`BpkDrawer should render correctly in the given target if renderTarget i
             </div>
           </header>
           <div
-            class="bpk-drawer__content"
+            class="bpk-drawer__content bpk-drawer__content--padded"
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/__snapshots__/BpkDrawer-test.js.snap:64]: Expect this class now for backwards compatible scenarios which don't opt out of padding

File: examples/bpk-component-drawer/examples.js [hunk 14]
```diff
@@ -262,10 +262,28 @@ const WithFullHeightContentExample = () => (
...
+  <DrawerContainer
+    title="Drawer title"
+    closeLabel="Close drawer"
+    buttonText="Open drawer"
+    getApplicationElement={() => document.getElementById('pagewrap')}
+    padded={false}
```
Comment (steviehailey-skyscanner) [inline examples/bpk-component-drawer/examples.js:271]: New prop (opt out of the default being true)

File: packages/bpk-component-drawer/src/BpkDrawerContent-test.tsx [hunk 15]
```diff
@@ -16,7 +16,7 @@
  * limitations under the License.
  */
 
-/* @flow strict */
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/BpkDrawerContent-test.tsx:19]: Convert to TS

File: packages/bpk-component-drawer/src/BpkDrawerContent-test.tsx [hunk 16]
```diff
@@ -63,6 +63,22 @@ describe('BpkDrawerContent', () => {
...
 
+  it('should render correctly when it has padded=true', () => {
+    const { asFragment } = render(
+      <BpkDrawerContent
+        id="my-drawer"
+        title="Drawer title"
+        onClose={jest.fn()}
+        onCloseAnimationComplete={jest.fn()}
+        dialogRef={jest.fn()}
+        padded
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/BpkDrawerContent-test.tsx:74]: Test new prop via sub-component (this doesn't have the true default val which the Bpk Component has)

File: packages/bpk-component-drawer/src/__snapshots__/BpkDrawerContent-test.tsx.snap [hunk 17]
```diff
@@ -165,6 +165,59 @@ exports[`BpkDrawerContent should render correctly when it has a contentClassName
 </DocumentFragment>
 `;
 
+exports[`BpkDrawerContent should render correctly when it has padded=true 1`] = `
```
Comment (steviehailey-skyscanner) [inline packages/bpk-component-drawer/src/__snapshots__/BpkDrawerContent-test.tsx.snap:168]: New scenario

**Review by chiefkes — APPROVED** [review]

File: packages/bpk-component-drawer/src/BpkDrawerContent.tsx [hunk 18]
```diff
@@ -34,42 +34,43 @@ import STYLES from './BpkDrawerContent.module.scss';
...
 
 type Props = {
-  children: Node,
-  dialogRef: () => mixed,
-  onCloseAnimationComplete: () => mixed,
-  onClose: () => mixed,
+  children: ReactNode,
+  dialogRef: () => React.RefObject<HTMLElement>,
+  onCloseAnimationComplete: () => void,
+  onClose: () => () => void
```
Comment (artemzhydkov96) [inline packages/bpk-component-drawer/src/BpkDrawerContent.tsx:40]: Does it mean that this function returns a function which returns void?

Comment (steviehailey-skyscanner) [reply packages/bpk-component-drawer/src/BpkDrawerContent.tsx]: This is a typo - good spot. Fix pushed. 
(It is just in the type def so not affecting the close behaviour and the tests are either in TS and using mocks and tolerant to it or not migrated from JS yet so didn't highlight it.)

**Review by artemzhydkov96 — APPROVED** [review]
LGTM

**Pull request merged**